src/main/java/org/elasticsearch/common/util/MultiDataPathUpgrader.java

```diff
@@ -0,0 +1,366 @@
+                        Files.createDirectories(targetDir);
+                        for (Path file : stream) {
+                            if (IndexWriter.WRITE_LOCK_NAME.equals(file.getFileName().toString())) {
+                                continue; // skip write.lock
+                            }
+                            logger.info("{} move file [{}] size: [{}]", shard, file.getFileName(), Files.size(file));
+                            final Path targetFile = targetDir.resolve(file.getFileName());
+                            Files.move(file, targetFile);
+                            movedFiles.add(targetFile);
+                        }
+                    }
+                }
+            }
+        }
+        if (movedFiles.isEmpty() == false) {
+            // fsync later it might be on disk already
+            logger.info("{} fsync files", shard);
+            for (Path moved : movedFiles) {
+                logger.info("{} syncing {}", shard, moved.getFileName());
+                IOUtils.fsync(moved, false);
+            }
+            logger.info("{} syncing directory {}", shard, targetDir);
+            IOUtils.fsync(targetDir, true);
+        }
+    }
+
+
+    /**
+     * Returns <code>true</code> iff the target path is one of the given paths.
+     */
+    private boolean isTargetPathConfigured(final Path[] paths, ShardPath targetPath) {
+        for (Path path : paths) {
+            if (path.equals(targetPath.getDataPath())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Runs an upgrade on all shards located under the given node environment
+     */
+    public static void upgradeMultiDataPath(NodeEnvironment nodeEnv) throws IOException {
+        MultiDataPathUpgrader upgrader = new MultiDataPathUpgrader(nodeEnv);
+        final Set<String> allIndices = nodeEnv.findAllIndices();
+
+        for (String index : allIndices) {
+            for (ShardId shardId : findAllShardIds(index, nodeEnv.indexPaths(new Index(index)))) {
+                try (ShardLock lock = nodeEnv.shardLock(shardId, 0)) {
+                    if (upgrader.needsUpgrading(shardId)) {
+                        final ShardPath shardPath = upgrader.pickShardPath(shardId);
+                        upgrader.upgrade(shardId, shardPath);
+                        // we have to check if the index path exists since we might
+                        // have only upgraded the shard state that is written under /indexname/shardid/_state
+                        // in the case we upgraded a dedicated index directory index
+                        if (Files.exists(shardPath.resolveIndex())) {
+                            upgrader.checkIndex(shardPath);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private static Set<ShardId> findAllShardIds(final String index, Path... locations) throws IOException {
```

Comment:
maybe it was intended to be passed to findAllShardsForIndex so it won't no rely on the folder name?